snuba/web/ast_rollout.py

```diff
@@ -0,0 +1,47 @@
+from typing import Optional
+import random
+
+from snuba import settings
+from snuba.state import get_config
+
+ROLLOUT_RATE_CONFIG = "ast_rollout_rate"
+KILLSWITCH_CONFIG = "ast_shutdown"
+
+
+def is_ast_rolled_out(dataset_name: str, referrer: Optional[str]) -> bool:
+    """
+    Applies some rules to decide whether we should run a query based
+    on the AST or on the legacy representation.
+
+    - if the emergency killswitch is on, then no rollout
+    - if a dataset is rolled out then use ast
+    - if the referrer is rolled out for a dataset then use ast
```

Comment:
You mean use the percentages defined for that dataset and referrer?

Reply:
Yes. I can rephrase.

Reply:
The other thing that I found unexpected is that I'd expect any more specific rule (dataset referrer rule) to apply first (before the dataset rule) however that isn't the case here.

Reply:
Well in practice the result is the same because neither the dataset nor the referrer-dataset rule can stop the flow. It is like if we had this condition:
```
if dataset_enabled OR referrer_dataset_enabled:
   return True
```

Anyway I will swap them

Reply:
Sorry, I phrased that poorly earlier, it wasn't quite what I meant. I guess I expected a rate that is specified in the dataset referrer rule to be always applied over what is in the dataset. Like if we have the following:
AST_REFERRER_ROLLOUT = {"transactions": {"something": 0}} 
AST_DATASET_ROLLOUT = {"transactions": 100}
...I'd expect is_ast_rolled_out("transactions", "something") to be False instead of True

Reply:
You are right, that makes sense. We could use it to selectively disable specific referrers. Making the change.